File: Problems/Algorithms/1642. Furthest Building You Can Reach/furthest_building.py
#!/usr/bin/env python3

import heapq as hq


class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        n = len(heights)
        if n == 1:
            return 0

        queue = []
        hq.heapify(queue)

        total, length = 0, 0
        for i in range(1, n):
            curr = heights[i] - heights[i-1]
            if curr > 0:
                if length < ladders:
                    length += 1
                    hq.heappush(queue, curr)
                else:
                    hq.heappush(queue, curr)
                    peak = hq.heappop(queue)
                    total += peak
                    if total > bricks:
                        return i-1

        return n-1

File: Problems/Algorithms/1642. Furthest Building You Can Reach/furthest-building.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        priority_queue<int> pq;

        int i = 0;
        while (i < heights.size() - 1) {
            int diff = heights[i+1] - heights[i];
            if (diff <= 0) {
                i++;
                continue;
            }

            bricks -= diff;
            pq.push(diff);

            if (bricks < 0) {
                bricks += pq.top();
                pq.pop();
                ladders--;
            }

            if (ladders < 0) {
                break;
            }

            i++;
        }

        return i;
    }
};
